File: c-interface/include/src/Application.cpp
#include "Astral.Canvas/Application.h"
#include "Application.hpp"

exportC const char *AstralCanvasApplication_GetApplicationName(AstralCanvasApplication ptr)
{
    return ((AstralCanvas::Application *)ptr)->appName.buffer;
}
exportC const char *AstralCanvasApplication_GetEngineName(AstralCanvasApplication ptr)
{
    return ((AstralCanvas::Application *)ptr)->engineName.buffer;
}
exportC float AstralCanvasApplication_GetFramesPerSecond(AstralCanvasApplication ptr)
{
    return ((AstralCanvas::Application *)ptr)->framesPerSecond;
}
exportC void AstralCanvasApplication_SetFramesPerSecond(AstralCanvasApplication ptr, float frames)
{
    ((AstralCanvas::Application *)ptr)->framesPerSecond = frames;
}
exportC void AstralCanvasApplication_AddWindow(AstralCanvasApplication ptr, i32 width, i32 height, bool resizeable)
{
    ((AstralCanvas::Application *)ptr)->AddWindow(width, height, resizeable);
}
exportC AstralCanvasApplication AstralCanvasApplication_Init(const char *appName, const char *engineName, u32 appVersion, u32 engineVersion, float framesPerSecond)
{
    SetDefaultAllocator(GetCAllocator());

    AstralCanvas::ApplicationInit(GetDefaultAllocator(), string(GetDefaultAllocator(), appName), string(GetDefaultAllocator(), engineName), appVersion, engineVersion, framesPerSecond);

    return (AstralCanvasApplication)AstralCanvas::GetAppInstance();
}
exportC AstralCanvasWindow AstralCanvasApplication_GetWindow(AstralCanvasApplication ptr, usize index)
{
    return (AstralCanvasWindow)((AstralCanvas::Application *)ptr)->windows.Get(index);
}

File: bindings/csharp/Application.cs
using System;
using System.Runtime.InteropServices;

namespace Astral.Canvas
{
    public class Application
    {
        public IntPtr handle;
        public int windowsCount;
        public Graphics graphicsDevice;

        public Application(string appName, string engineName, uint appVersion, uint engineVersion, float fps)
        {
            handle = AstralCanvas.Application_Init(appName, engineName, appVersion, engineVersion, fps);
            graphicsDevice = new Graphics(this);
        }

        public string applicationName
        {
            get
            {
                return Marshal.PtrToStringAnsi(AstralCanvas.Application_GetApplicationName(handle));
            }
        }
        public string engineName
        {
            get
            {
                return Marshal.PtrToStringAnsi(AstralCanvas.Application_GetEngineName(handle));
            }
        }
        public float framesPerSecond
        {
            get
            {
                return AstralCanvas.Application_GetFramesPerSecond(handle);
            }
            set
            {
                AstralCanvas.Application_SetFramesPerSecond(handle, value);
            }
        }
        public void AddWindow(int width, int height, bool resizeable)
        {
            AstralCanvas.Application_AddWindow(handle, width, height, resizeable);
            windowsCount += 1;
        }
        public Window GetWindow(int index)
        {
            if (index >= windowsCount)
            {
                throw new IndexOutOfRangeException();
            }
            return new Window(AstralCanvas.Application_GetWindow(handle, (uint)index));
        }

        public void Run(AstralCanvas.UpdateFunction onUpdate, AstralCanvas.UpdateFunction onDraw, AstralCanvas.InitFunction onProgramInitialize, AstralCanvas.DeinitFunction onProgramEnd)
        {
            AstralCanvas.Application_Run(handle, onUpdate, onDraw, onProgramInitialize, onProgramEnd);
        }
    }
}
